src/gsAssembler/gsPatchRule.hpp

```diff
@@ -0,0 +1,402 @@
+    // This number can be different per element
+    // we also compute the total number of points in the tensor product (size)
+    index_t size = 1;
+    std::vector<gsVector<T> > elNodes(m_dim), elWeights(m_dim);
+    index_t k = 0; // counter per direction d
+    for (size_t d = 0; d!=m_dim; d++)
+    {
+        elNodes[d].resize(m_nodes[d].size());
+        elWeights[d].resize(m_weights[d].size());
+        for (auto it = m_maps[d].lower_bound(lower[d]); it!= ( upper[d]==m_end[d] ? m_maps[d].upper_bound(upper[d]) : m_maps[d].lower_bound(upper[d]) ); it++, k++) // lower_bound = geq, upper_bound= greather than
+        {
+            elNodes[d].at(k) = it->first;
+            elWeights[d].at(k) = it->second;
+        }
+        elNodes[d].conservativeResize(k);
+        elWeights[d].conservativeResize(k);
+        size *= elNodes[d].size(); // compute tensor size
+        k = 0; // reset counter
+    }
+
+    // This could work. However, in cases when there are elements which have no quadPoint, it goes wrong
+    // this->computeTensorProductRule_into(elNodes,elWeights,nodes,weights);
+
+    // initialize the number of nodes and weights
+    nodes.resize(m_dim,size);
+    weights.resize(size);
+    if (size==0)
+        return;
+
+    // Now we fill the matrix with the points and we construct the tensor product (according to the scheme on top)
+    gsMatrix<T> tmpNodes, tmpWeights;
+    gsVector<T> ones;
+    tmpNodes = elNodes[0].transpose();
+    tmpWeights = elWeights[0].transpose();
+    size = 1;
+    for (size_t d = 1; d!=m_dim; d++)
+    {
+        nodes.block( 0, 0, d, tmpNodes.cols()*elNodes[d].size() ) = tmpNodes.replicate(1,elNodes[d].size());
+
+        ones.setOnes(tmpNodes.cols());
+        for (index_t k = 0; k != elNodes[d].size(); k++)
+        {
+            nodes.block(d,k*ones.size(),1,ones.size()) = ones.transpose()*elNodes[d].at(k);
+            weights.segment(k*ones.size(),ones.size()) = (ones.transpose()*elWeights[d].at(k)).cwiseProduct(tmpWeights);
+        }
+
+        tmpWeights.transpose() = weights.segment(0,tmpWeights.cols()*elWeights[d].size());
+        tmpNodes = nodes.block( 0, 0, d+1, tmpNodes.cols()*elNodes[d].size() );
+    }
+};
+
+template<class T> void
+gsPatchRule<T>::mapTo( T startVal, T endVal,
+                      gsMatrix<T> & nodes, gsVector<T> & weights ) const
+{
+    GISMO_ASSERT( 1 == m_nodes.size(), "Inconsistent quadrature mapping (dimension != 1)");
+
+    nodes.resize(1,m_nodes[0].size());
+    weights.resize(m_weights[0].size());
+    index_t k=0;
+    for (auto it = m_maps[0].lower_bound(startVal); it!=m_maps[0].upper_bound(endVal); it++, k++) // lower_bound = geq, upper_bound= greather than
+    {
+        nodes.at(k) = it->first;
+        weights.at(k) = it->second;
+    }
+    nodes.conservativeResize(1,k);
+    weights.conservativeResize(k);
+}
+
+/*
+// THIS IS AN IMPLEMENTATION FOR STRICTLY 2D DOMAINS!
```

Comment:
maybe you can remove the commented 2d code by now

Reply:
Done